pyartcd/pyartcd/pipelines/gen_assembly.py

```diff
@@ -0,0 +1,233 @@
+        self.nightlies = nightlies
+        self.allow_pending = allow_pending
+        self.allow_rejected = allow_rejected
+        self.allow_inconsistency = allow_inconsistency
+        self.custom = custom
+        self.arches = arches
+        self.in_flight = in_flight
+        self.previous_list = previous_list
+        self.auto_previous = auto_previous
+        self._logger = logger or runtime.logger
+        self._slack_client = self.runtime.new_slack_client()
+        self._working_dir = self.runtime.working_dir.absolute()
+
+        # determines OCP version
+        match = re.fullmatch(r"openshift-(\d+).(\d+)", group)
+        if not match:
+            raise ValueError(f"Invalid group name: {group}")
+        self._ocp_version = (int(match[1]), int(match[2]))
+
+        # sets environment variables for Doozer
+        self._doozer_env_vars = os.environ.copy()
+        self._doozer_env_vars["DOOZER_WORKING_DIR"] = str(self._working_dir / "doozer-working")
+
+        if not ocp_build_data_url:
+            ocp_build_data_url = self.runtime.config.get("build_config", {}).get("ocp_build_data_url")
+        if ocp_build_data_url:
+            self._doozer_env_vars["DOOZER_DATA_PATH"] = ocp_build_data_url
+
+    async def run(self):
+        self._slack_client.bind_channel(self.group)
+        slack_response = await self._slack_client.say(f":construction: Generating assembly definition {self.assembly} :construction:")
+        slack_thread = slack_response["message"]["ts"]
+        try:
+            if self.arches and self.custom:
+                raise ValueError("Customizing arches can only be used with custom assemblies.")
+
+            self._logger.info("Getting nightlies from Release Controllers...")
+            candidate_nightlies = await self._get_nightlies()
+            self._logger.info("Generating assembly definition...")
+            assembly_definition = await self._gen_assembly_from_releases(candidate_nightlies)
+            out = StringIO()
+            yaml.dump(assembly_definition, out)
+            self._logger.info("Generated assembly definition:\n%s", out.getvalue())
+
+            # Create a PR
+            def _merge(a, b):
+                if not isinstance(a, dict) or not isinstance(b, dict):
+                    return a
+                c: OrderedDict = b.copy()
+                for k, v in a.items():
+                    c[k] = _merge(v, b.get(k))
+                    if k not in b:
+                        # move new entry to the beginning
+                        c.move_to_end(k, last=False)
+                return c
+            pr = await self._create_or_update_pull_request(lambda releases_yaml: _merge(assembly_definition, releases_yaml))
+
+            # Sends a slack message
+            message = f"Hi @release-artists , please review assembly definition for {self.assembly}: {pr.html_url}"
+            await self._slack_client.say(message, slack_thread)
+        except Exception as err:
```

Comment:
Do we expect other types of exceptions apart from ChildProcessError?

Reply:
This is a catch-all statement to capture whatever exception and send it to Slack. No need to worry about the concrete type here.